homeassistant/components/ssdp/__init__.py

```diff
@@ -69,108 +80,134 @@ def __init__(self, hass, integration_matchers):
         """Initialize class."""
         self.hass = hass
         self.seen = set()
-        self._entries = []
         self._integration_matchers = integration_matchers
-        self._description_cache = {}
+        self._cancel_scan = None
+        self._ssdp_listeners = []
+        self._callbacks = []
+        self.flow_dispatcher: FlowDispatcher | None = None
+        self.description_manager: DescriptionManager | None = None
 
-    async def _on_ssdp_response(self, data: Mapping[str, Any]) -> None:
+    async def _async_on_ssdp_response(self, data: Mapping[str, Any]) -> None:
         """Process an ssdp response."""
-        self.async_store_entry(
+        await self._async_process_entry(
             ssdp.UPNPEntry({key.lower(): item for key, item in data.items()})
         )
 
-    @callback
-    def async_store_entry(self, entry):
-        """Save an entry for later processing."""
-        self._entries.append(entry)
-
-    async def async_scan(self, _):
-        """Scan for new entries."""
-
-        await async_search(async_callback=self._on_ssdp_response)
-        await self._process_entries()
-
-        # We clear the cache after each run. We track discovered entries
-        # so will never need a description twice.
-        self._description_cache.clear()
-        self._entries.clear()
-
-    async def _process_entries(self):
-        """Process SSDP entries."""
-        entries_to_process = []
-        unseen_locations = set()
-
-        for entry in self._entries:
-            key = (entry.st, entry.location)
-
-            if key in self.seen:
+    @core_callback
+    def async_register_callback(self, ssdp_callback, match_dict=None):
+        """Register a callback."""
+        if match_dict is None:
+            match_dict = {}
+
+        callback_entry = (ssdp_callback, match_dict)
+        self._callbacks.append(callback_entry)
+
+        @core_callback
+        def _async_remove_callback():
+            self._callbacks.remove(callback_entry)
+
+        return _async_remove_callback
+
+    @core_callback
+    def async_stop(self, *_):
+        """Stop the scanner."""
+        self._cancel_scan()
+        for listener in self._ssdp_listeners:
+            listener.async_stop()
+        self._ssdp_listeners = []
+
+    async def _async_build_source_set(self):
+        """Build the list of ssdp sources."""
+        adapters = await network.async_get_adapters(self.hass)
+        sources = set()
+        if _async_use_default_interface(adapters):
+            sources.add(IPv4Address("0.0.0.0"))
+            return sources
+
+        for adapter in adapters:
+            if not adapter["enabled"]:
                 continue
+            if adapter["ipv4"]:
+                ipv4 = adapter["ipv4"][0]
+                sources.add(IPv4Address(ipv4["address"]))
+            if not adapter["ipv6"]:
```

Comment:
I had to refactor this because I couldn't use the python 3.9 scope.  I forgot to remove the continue. Thanks!

Fixed in ef6a2042a768f493eec7b53db363c923e60bacd4